WMF Framework/EventPlatformClientLibrary.swift

```diff
@@ -597,98 +660,59 @@ public class EPC: NSObject {
      *   `MWKLanguageLinkController.sharedInstance().appLanguage.siteURL().host`
      */
-    @objc public func submit(stream: String, data: [String: NSCoding], domain: String? = nil) -> Void {
-        guard self.sharingUsageData else {
-            return
-        }
-
-        /*
-         * EventGate needs to know which version of the schema to validate against
-         */
-        guard data.keys.contains("$schema") else {
-            DDLogError("EPC: Event data is missing the required '$schema' field")
-            return
-        }
-
-        var meta: [String: NSCoding] = data["meta"] as? [String: NSCoding] ?? [:]
-
-        if let domain = domain {
-            meta["domain"] = domain as NSCoding
-        }
-
-        var data = data
-        data["meta"] = meta as NSCoding
-
-        /*
-         * The top-level field `client_dt` is for recording the time the event
-         * was generated. EventGate sets `meta.dt` during ingestion, so for
-         * analytics events that field is used as "timestamp of reception" and
-         * is used for partitioning the events in the database. See Phab:T240460
-         * for more information.
-         */
-        if !data.keys.contains("client_dt") {
-            data["client_dt"] = iso8601Formatter.string(from: Date()) as NSCoding
-        }
-
-        /*
-         * Generated events have the session ID attached to them before stream
-         * config is available (in case they're generated offline) and before
-         * they're cc'd to any other streams (once config is available).
-         */
-        if !data.keys.contains("app_session_id") {
-            data["app_session_id"] = sessionID as NSCoding
-        }
-
-        #if DEBUG
-        do {
-            let jsonString = try data.toPrettyPrintJSONString()
-            DDLogDebug("EPC: Event logged to stream '\(stream)':\n\(jsonString)")
-        } catch let error {
-            DDLogError("EPC: \(error.localizedDescription)")
-        }
-        #endif
-
-        guard let streamConfigs = streamConfigurations else {
-            let event: (stream: String, data: [String: NSCoding], domain: String?) = (stream: stream, data: data, domain: domain)
-            appendEventToInputBuffer(event)
-            return
+    public func submit<E: EventInterface>(stream: Stream, event: E, domain: String? = nil, date: Date? = nil) {
+        encodeQueue.async {
+            self._submit(stream: stream, event: event, domain: domain, date: date)
         }
+    }
 
-        if !(streamConfigs.keys.contains(stream)) {
-            DDLogDebug("EPC: Event submitted to '\(stream)' but only the following streams are configured: \(streamConfigs.keys.joined(separator: ", "))")
-            return
-        }
-        
-        if !inSample(stream: stream) {
+    /// Private, synchronous version of `submit`.
+    private func _submit<E: EventInterface>(stream: Stream, event: E, domain: String? = nil, date: Date? = nil){
```

Comment:
This might be more of a comment with the original PR and less with the refactor, but is there any need anymore for the `date` parameter? I don't see how it could be used, even if events are queued in the inputBuffer first and posted after the config fetches (which I think is how it was originally going to be used).

Reply:
My understanding of the original PR was that it's for providing a custom `client_dt` but if we don't plan to utilize that, it could be removed.

Reply:
`client_dt` needs to be the timestamp of when the event was originally submitted. The original PR checks for its presence in the event data and if it's not already there (meaning the event is being submitted for the first time and not re-submitted after stream config is available) it sets it. I'm not entirely sure where this is set in this PR.